pkg/ko/publish/daemon.go

```diff
@@ -47,14 +48,23 @@ func (d *demon) Publish(img v1.Image, s string) (name.Reference, error) {
 	if err != nil {
 		return nil, err
 	}
+
+	for _, tagName := range d.tags {
+		tag, err := name.NewTag(fmt.Sprintf("%s/%s:%s", LocalDomain, d.namer(s), tagName), name.WeakValidation)
+		if err != nil {
+			return nil, err
+		}
+		log.Printf("Loading %v", tag)
+		if _, err := daemon.Write(tag, img); err != nil {
+			return nil, err
+		}
+		log.Printf("Loaded %v", tag)
+	}
+
 	tag, err := name.NewTag(fmt.Sprintf("%s/%s:%s", LocalDomain, d.namer(s), h.Hex), name.WeakValidation)
```

Comment:
I think we need to still load this one, since we'll be referencing it later. You could create a new list of tags that is `d.tags` + this tag, and use that for the above loop.

Reply:
Actually we don't. Any image that is published with a tag can be referenced by it's digest, so we don't need to upload it via the digest.

Reply:
Omg. I just typed up a huge response but unresolving the conversation made it go away. One second, I'll try to reproduce it.

Reply:
I'm pretty sure we do still need this tag.

When using `--local`, we don't actually publish the images to a registry, we just sideload them into the docker daemon. Docker doesn't calculate these image digests until you push the image (it's expensive to calculate the digest because you have to gzip all the layers), so we can't reference them by their digests. Instead, we _tag_ these images with the digest, e.g.:

```
ko.local/webhook-37fa64b5b216d563332fd3210a121e52:5d887242fa90e49a31bb1f7caf9f2194de3890251acb65610f98fe60dd601cb1
```

instead of:

```
ko.local/webhook-37fa64b5b216d563332fd3210a121e52@sha256:5d887242fa90e49a31bb1f7caf9f2194de3890251acb65610f98fe60dd601cb1
```

It's a bit of a mess, but it's the only way to get minikube to work without a registry (as far as I know, we can't coerce docker into calculating a digest for us).

I don't have minikube installed, but you can see here that after running `ko resolve --tags "foo" -L -f config/` with your changes, docker doesn't know the digests for those images:
```
$ docker images --digests
REPOSITORY                                             TAG                 DIGEST                                                                    IMAGE ID            CREATED             SIZE
ko.local/webhook-37fa64b5b216d563332fd3210a121e52      foo                 <none>                                                                    4a98e85efff9        49 years ago        53.4MB
ko.local/activator-2d008c4c36689d255523e5fa210c0bec    foo                 <none>                                                                    9e529f7fe80a        49 years ago        61.2MB
ko.local/autoscaler-b41733a3e5694e8fc64777bdfd8e791a   foo                 <none>                                                                    f2315e73e8a3        49 years ago        62.1MB
ko.local/controller-43f0364ab2f6dab17267e80a1f6a4adc   foo                 <none>                                                                    48f90164b572        49 years ago        64.3MB
ko.local/queue-7204c16e44715cd30f78443fb99e0f58        foo                 <none>                                                                    dd1a04b833ea        49 years ago        50.9MB
```

Reply:
Wow, that is a shame. Didn't know about this about the docker daemon, found some github issues from moby/docker that exposes this problem. I tested this on gcr, so i always end up getting an image with digest.

I'll add this tomorrow. So we actually only need to do this on the local daemon. Remote writes doens't need, right?

Thanks for clarifying this.